### PR TITLE
Include Pinned Resources in User Preferences

### LINKED ISSUE #32008: Add pinned resource IDs to user prefernces

### PR DESCRIPTION
**Pull request by avatus**
Closes https://github.com/gravitational/teleport/issues/32008
In support of https://github.com/gravitational/teleport/issues/30418
This will add Pinned Resources to the user preferences object. The shape on web is:
```js
const pinnedResources = {
    "clustername1": ["node1", "node2", "node3"],
    "clustername2": ["node4", "node5", "node6"],
}
```
We will use this to get a list of resource IDs to display which resources are pinned in the unified resource view. 
Screenshot 2023-09-14 at 5 20 09 PM

node: two PRs will be be opened that rely on this. The fetching of the actual resource based on the users preferences, and updating which resources are saved via the Pin button. Both of those will be opened shortly after this one.

It can be tested by inspecting the user preferences local storage object after logging in (it will be empty but there!)

### CODE REVIEW

File: lib/services/local/userpreferences_test.go [hunk 1]
```diff
@@ -141,6 +164,7 @@ func TestUserPreferencesCRUD2(t *testing.T) {
 					Onboard: &userpreferencesv1.OnboardUserPreferences{
 						PreferredResources: []userpreferencesv1.Resource{userpreferencesv1.Resource_RESOURCE_KUBERNETES},
 					},
+					PinnedResources: pinned,
```
Comment (kimlisa) [inline lib/services/local/userpreferences_test.go:167]: this test case is for "updating all settings at once" but since we already updated to this `pinned` previously, it's not really meeting the test case?

File: web/packages/teleport/src/services/userPreferences/types.ts [hunk 2]
```diff
@@ -36,10 +36,13 @@ export type OnboardUserPreferences = {
   preferredResources: ClusterResource[];
 };
 
+export type PinnedResourcesUserPreferences = Record<string, string[]>;
```
Comment (kimlisa) [inline web/packages/teleport/src/services/userPreferences/types.ts:39]: i'd add a comment that the key is the clusterid and values are uuid's

File: lib/web/userpreferences.go [hunk 3]
```diff
@@ -91,17 +93,40 @@ func (h *Handler) updateUserPreferences(_ http.ResponseWriter, r *http.Request,
 	return OK(), nil
 }
 
+func convertToPinnedResources(pinnedResources map[string][]string) *userpreferencesv1.PinnedResourcesUserPreferences {
```
Comment (kimlisa) [inline lib/web/userpreferences.go:96]:
```suggestion
func convertToPinnedResourcesUserPreferencesProto(pinnedResources map[string][]string) *userpreferencesv1.PinnedResourcesUserPreferences {
```

File: lib/web/userpreferences.go [hunk 4]
```diff
@@ -91,17 +93,40 @@ func (h *Handler) updateUserPreferences(_ http.ResponseWriter, r *http.Request,
...
+	pinnedResourcesProto := make(map[string]*userpreferencesv1.ClusterPinnedResources)
```
Comment (kimlisa) [inline lib/web/userpreferences.go:97]: specify size?

File: lib/web/userpreferences.go [hunk 5]
```diff
@@ -91,17 +93,40 @@ func (h *Handler) updateUserPreferences(_ http.ResponseWriter, r *http.Request,
...
+		Theme:           resp.Theme,
+		Onboard:         onboardUserPreferencesResponse(resp.Onboard),
+		PinnedResources: pinnedResourcesUserPreferencesResponse(resp.PinnedResources),
 	}
 
 	return jsonResp
 }
 
+func pinnedResourcesUserPreferencesResponse(resp *userpreferencesv1.PinnedResourcesUserPreferences) map[string][]string {
+	pinnedResources := make(map[string][]string)
```
Comment (kimlisa) [inline lib/web/userpreferences.go:123]: specify size?

File: lib/services/local/userpreferences.go [hunk 6]
```diff
@@ -177,6 +178,14 @@ func overwriteValues(dst, src protoreflect.ProtoMessage) error {
...
 	src.Range(func(fd protoreflect.FieldDescriptor, v protoreflect.Value) bool {
 		switch {
+		case fd.IsMap():
+			srcMap := v.Map()
+			dstMap := dst.Mutable(fd).Map()
+
+			srcMap.Range(func(k protoreflect.MapKey, vv protoreflect.Value) bool {
+				dstMap.Set(k, vv)
+				return true
+			})
```
Comment (kimlisa) [inline lib/services/local/userpreferences.go:188]: just curious why this and the function is required?

File: lib/services/local/userpreferences_test.go (hunk 1 above)
Comment (avatus) [reply lib/services/local/userpreferences_test.go:167]: Each test creates a new user preferences service so any upserting to user prefs is unique by test case. Unless I'm mistaken

File: lib/services/local/userpreferences.go (hunk 6 above)
Comment (avatus) [reply lib/services/local/userpreferences.go:188]: I was running into an issue when updating user preferences because its a "map" type and not a "message" type. So I had to first check if the incoming value to the method was a map in order to update it differently

**Review by kimlisa — APPROVED** [review]

File: api/proto/teleport/userpreferences/v1/pinned_resources.proto [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+message ClusterPinnedResources {
+  // resource_ids is a list of resource ids
+  repeated string resource_ids = 1;
+}
+
+// PinnedResourcesUserPreferences is a collection of resource IDs that will be
+// displayed in the user's pinned resources tab in the Web UI
+message PinnedResourcesUserPreferences {
+  // pinned_resources is a map of resource IDs pinned per cluster
+  map<string, ClusterPinnedResources> pinned_resources = 1;
```
[the remark this answers is not in the document]
Comment (avatus) [reply api/proto/teleport/userpreferences/v1/pinned_resources.proto:31]: > Should pinned resources be stored per cluster rather than having them all in the root cluster?

I'm not sure yet. It seems reasonable enough to allow it but I haven't done any super thinking on it yet.

> Is there an upper bound to the number of pinned resources? Is it per cluster or total? 

There isn't an upperbound programmed in yet, but we'd want to stay under the max size of an item, lets say in dynamodb, which is 400kb. if we have a node id like "123123-123123123-123123123/node", well over 10k could fit and still have plenty of room for more (and other user preferences).

I don't think 10k pinned resources across all clusters is a realistic outcome in terms of what people actually pin, but we should put some safeguards up regardless. and also it comes into play with my next point. 

> How do pinned resources get cleaned up? Is it manual? What happens if a leaf cluster is removed entirely?

This will require some extra thinking. My first inclination is that when a pinned resources request is made, we can track how many resources "couldn't be found" from their pinned list, and then just remove them. But I don't think thats a good place to do it as I wouldn't want to add extra latency to the request. there should be some way to clean them up however, because if that was left to grow unbounded then we could run into the issue above.

> What will be displayed to a user if a pinned resource is removed? 

If they pinned something that gets removed or they lose access to it, it just wouldn't be in shown. a "pinned resource" is really just an extra filter with specific IDs. you can do the same with an advanced query, so the same UX if they opened the normal tab and a resource was gone. it just wouldn't be there. 

> Is the resource id enough information for a user to discern which resource they pinned?

This isn't super relevant yet since they won't see something that doesn't exist. However, I could envision a future where there is a 'user preferences' page that could show the IDs they have pinned (existing and not) and be able to edit it there. Then maybe ID wouldn't be enough, but ID + if it exists or not (with info from existing resources surrounding it) might be enough. That bridge is very far down the road tho.


To summarize, the main points that should probably be solved now-ish or at least documented are the cleanup phase and the max size. I will try to get something written up about this

[the remark this answers is not in the document]
Comment (fspmarshall) [reply api/proto/teleport/userpreferences/v1/pinned_resources.proto:31]: Based on the above, I'd lean toward storing the pins per-cluster rather than storing all pins in root as is currently done.  If we store pins per cluster, doing automatic cleanup as a periodic won't be too difficult.  If, on the other hand, we do all pinning in the root that makes things _very_ complicated.  Cross-cluster communication is non-trivial.

Also note that a resource disappearing temporarily doesn't necessarily mean that it's gone.  An automatic cleanup routine should probably track timestamps and only clean up resource that remain missing for some minimum amount of time + number of checks.

IMO wether or not this PR includes a cleanup mechanism, we should have a reasonably clear idea of how the cleanup mechanism will function prior to merging it.

[the remark this answers is not in the document]
Comment (zmb3) [reply api/proto/teleport/userpreferences/v1/pinned_resources.proto:31]: I don't think it's a bad idea to put an upper limit on the number of pinned resources. Pinning resources becomes less helpful the more you pin, so I'd expect people won't pin too many. Maybe we set a hard cap to something like 300 or 500?

As far as cleanup goes, I think for now it's okay if your preferences reference a resource that no longer exists - they just won't show up in the listing.

We can work with @roraback to get some designs for how you might remove them.

**Comment by avatus** [comment]
Updated to store cluster preferences per cluster. Still need to add the pinned resources cap

File: lib/services/local/userpreferences.go [hunk 8]
```diff
@@ -83,25 +83,7 @@ func (u *UserPreferencesService) UpsertUserPreferences(ctx context.Context, user
...
-	if prefs.ClusterPreferences != nil {
-		preferences.ClusterPreferences = prefs.ClusterPreferences
-	} else {
-		preferences.Assist = prefs.Assist
-		preferences.Onboard = prefs.Onboard
-		preferences.Theme = prefs.Theme
-	}
-
-	item, err := createBackendItem(username, preferences)
+	item, err := createBackendItem(username, prefs)
```
Comment (avatus) [inline lib/services/local/userpreferences.go]: This was removed because originally, we'd only be sending small chunks of the user preferences to update. Now we send the entire thing (before this PR) so we can just completely overwrite the prefs. Using a range to go through what was sent in the protomessage would skip empty values and therefor not allow us to "empty" our pinned resources. Now that we send the entire thing, it works as expected.

**Review by rosstimothy — APPROVED** [review]
